### PR TITLE
Discovery

### PR DESCRIPTION
**Pull request by theJenix**
New discovery process, complete with multi select dialogs and support for a remote fan to initiate "find new fans".  There is still a bug on the "fan" side when the app has been running a long time, and there's still a question about how we integrate it into the main app (I'm thinking another service maybe), but this is tested working and can be pulled in while we discuss that other stuff.

### CODE REVIEW

File: fanClub/src/com/lastcrusade/fanclub/HostActivity.java [hunk 1]
```diff
@@ -15,34 +14,49 @@
...
+import com.lastcrusade.fanclub.net.message.FoundFan;
+import com.lastcrusade.fanclub.net.message.FoundFansMessage;
 import com.lastcrusade.fanclub.net.message.StringMessage;
 import com.lastcrusade.fanclub.util.BluetoothUtils;
 import com.lastcrusade.fanclub.util.Toaster;
 
+/**
+ * 
+ * 
+ * @author Jesse Rosalia
```
Comment (bryansills) [inline fanClub/src/com/lastcrusade/fanclub/HostActivity.java:44]: now you are just taunting us

File: fanClub-test/src/com/lastcrusade/fanclub/net/message/ConnectFansMessageTest.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        List<String> addresses = new ArrayList<String>();
+        addresses.add("00:11:22:33:44:55:66");
+        addresses.add("00:11:22:33:44:55:67");
+        addresses.add("00:11:22:33:44:55:68");
+        addresses.add("00:11:22:33:44:55:69");
+
+        ConnectFansMessage oldMessage = new ConnectFansMessage(addresses);
+        ConnectFansMessage newMessage = super.testSerializeMessage(oldMessage);
+        //verify that the FoundFan objects are equal and in the same order.
+        for (int ii = 0; ii < addresses.size(); ii++) {
```
Comment (ejohnson44) [inline fanClub-test/src/com/lastcrusade/fanclub/net/message/ConnectFansMessageTest.java:25]: Why ii instead of the normal i or index?

Comment (theJenix) [inline fanClub-test/src/com/lastcrusade/fanclub/net/message/ConnectFansMessageTest.java:25]: A habit impressed upon me by my first boss.  ii is much easier to search for within a file than i, as very few words are spelled with 'ii'.

File: fanClub/src/com/lastcrusade/fanclub/HostActivity.java (hunk 1 above)
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/HostActivity.java:44]: :laughing:

File: fanClub/src/com/lastcrusade/fanclub/components/MultiSelectListDialog.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        this.onClickListener = new IOnDialogItemClickListener<T>() {
+
+            @Override
+            public void onItemClick(T item) {
+                //NO OP
+            }
+        };
+    }
+
+    public MultiSelectListDialog<T> setFormatter(
```
Comment (bryansills) [inline fanClub/src/com/lastcrusade/fanclub/components/MultiSelectListDialog.java:56]: Why are you returning the object with a setter?

Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/components/MultiSelectListDialog.java:56]: I used the builder pattern with this dialog..the reason you return this from all of those methods is so you can chain methods together.

File: fanClub/src/com/lastcrusade/fanclub/net/BluetoothDiscoveryHandler.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        this.activity.sendBroadcast(intent);
+    }
+
+    /**
+     * Call to hold a discovered device for selection by the user.
+     * 
+     * @param device
+     */
+    public void onDiscoveryFound(BluetoothDevice device) {
+        Log.w(TAG,
```
Comment (bryansills) [inline fanClub/src/com/lastcrusade/fanclub/net/BluetoothDiscoveryHandler.java:74]: Couldn't we just use BluetoothDeviceDialogFormatter.format()?

File: fanClub-test/src/com/lastcrusade/fanclub/net/message/ConnectFansMessageTest.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+public class ConnectFansMessageTest extends SerializationTest<ConnectFansMessage> {
+
+    @Test
+    public void testSerializeFoundFansMessage() throws IOException {
+        //NOTE: no fields to check, and the base class will ensure we create the right class.
+        List<String> addresses = new ArrayList<String>();
+        addresses.add("00:11:22:33:44:55:66");
+        addresses.add("00:11:22:33:44:55:67");
+        addresses.add("00:11:22:33:44:55:68");
+        addresses.add("00:11:22:33:44:55:69");
```
Comment (reidbaker) [inline fanClub-test/src/com/lastcrusade/fanclub/net/message/ConnectFansMessageTest.java:20]: hehe 69

File: fanClub/lint.xml [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (reidbaker) [inline fanClub/lint.xml]: Just out of curiosity what is this file for?

File: fanClub/src/com/lastcrusade/fanclub/FanActivity.java [hunk 7]
```diff
@@ -88,6 +97,30 @@ public void onClick(View v) {
                 onHelloButtonClicked();
```
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/FanActivity.java:97]: Does this method exist anymore?

File: fanClub/src/com/lastcrusade/fanclub/FanActivity.java [hunk 8]
```diff
@@ -88,6 +97,30 @@ public void onClick(View v) {
...
+
+    private void registerMessageHandlers() {
+        this.messageDispatch = new MessageThreadMessageDispatch();
+        this.messageDispatch.registerHandler(StringMessage.class, new IMessageHandler<StringMessage>() {
+
+            @Override
+            public void handleMessage(int messageNo,
+                    StringMessage message, String fromAddr) {
+                StringMessage sm = (StringMessage)message;
+                Toaster.iToast(FanActivity.this, sm.getString());
```
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/FanActivity.java:112]: Is this toast for testing or will it be showing the user important info?

File: fanClub/src/com/lastcrusade/fanclub/HostActivity.java (hunk 1 above)
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/HostActivity.java:44]: You mean like this kind of taunting? 
http://www.youtube.com/watch?feature=player_detailpage&v=9V7zbWNznbs#t=156s

File: fanClub/src/com/lastcrusade/fanclub/HostActivity.java [hunk 9]
```diff
@@ -88,48 +103,106 @@ public void onClick(View v) {
...
+                    textBuffer.append(sm.getString());
+                }
+            }
+            
+        });
+        this.messageDispatch.registerHandler(FindNewFansMessage.class, new IMessageHandler<FindNewFansMessage>() {
+
+            @Override
+            public void handleMessage(int messageNo,
+                    FindNewFansMessage message, String fromAddr) {
```
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/HostActivity.java:132]: `fromAddr` is a String but what would be an example of what is in fromAddr?
Is it like a mac or bluetooth name?

File: fanClub/src/com/lastcrusade/fanclub/HostActivity.java [hunk 10]
```diff
@@ -144,12 +217,80 @@ protected void onHelloButtonClicked() {
         }
     }
 
+    /**
+     * Send a message to all connected clients
+     * 
+     * @param sm
```
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/HostActivity.java]: If we are going to have docstrings lets say what sm is haha

File: fanClub/src/com/lastcrusade/fanclub/HostActivity.java [hunk 11]
```diff
@@ -158,111 +299,46 @@ private void broadcastMessage(StringMessage sm) {
...
-        List<UUID> supportedUuids = BluetoothUtils.getUuidsForDevice(device);
-        UUID uuid = UUID.fromString(this.getString(R.string.app_uuid));
-        if (uuid == null) {
-            throw new RuntimeException("string.app_uuid is unset.");
+    private void handleFindNewFansMessage(final String remoteAddr) {
+        Toaster.iToast(this, R.string.finding_new_fans);
+
+        BluetoothAdapter adapter = BluetoothAdapter.getDefaultAdapter();
+        if (adapter == null) {
+            Toaster.eToast(this, "Bluetooth adapter is null");
```
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/HostActivity.java]: We might want to make this a string in strings.xml.
If it gets shown to the user it should be in an xml file. In fact we may want to have an error.xml with error messages.

File: fanClub/src/com/lastcrusade/fanclub/HostActivity.java [hunk 12]
```diff
@@ -158,111 +299,46 @@ private void broadcastMessage(StringMessage sm) {
...
-            synchronized (msgMutex) {
-                if (this.message.length() > 0) {
-                    this.message.append("\n");
-                } else {
-                    startDelayedDisplayMessage();
-                }
-                this.message.append(sm.getString());
-            }
+        if (found == null) {
+            Toaster.eToast(this, "Unknown remote device.");
```
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/HostActivity.java]: We might want to make this a string in strings.xml.
If it gets shown to the user it should be in an xml file. In fact we may want to have an error.xml with error messages.

File: fanClub/src/com/lastcrusade/fanclub/components/MultiSelectListDialog.java (hunk 3 above)
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/components/MultiSelectListDialog.java:56]: Is there a place in code that uses the chaining you mention?

Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/components/MultiSelectListDialog.java:56]: HostActivity#onDiscoveredDevices

File: fanClub/src/com/lastcrusade/fanclub/net/MessageThreadMessageDispatch.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.Map;
+
+import android.os.Handler;
+import android.os.Message;
+import android.util.Log;
+
+import com.lastcrusade.fanclub.net.message.IMessage;
+
+/**
+ * An Android Handler used as a fanClub network message dispatch.  This will receive Android Message objects that represent
```
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/net/MessageThreadMessageDispatch.java]: Long line

File: fanClub/src/com/lastcrusade/fanclub/net/MessageThreadMessageDispatch.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+ */
+public class MessageThreadMessageDispatch extends Handler {
+
+    public interface IMessageHandler<T> {
+        public void handleMessage(int messageNo, T message, String fromAddr);
+    }
+
+    private static final String TAG = "MessageThreadMessageDispatch";
+
+    private Map<Class<? extends IMessage>, IMessageHandler<? extends IMessage>> dispatchMap =
```
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/net/MessageThreadMessageDispatch.java:41]: Well this line isn' scary. It just needs some more punctuation.

File: fanClub/src/com/lastcrusade/fanclub/net/message/FindNewFansMessage.java [hunk 15]
```diff
@@ -8,9 +8,11 @@
 
     @Override
     public void deserialize(InputStream input) throws IOException {
```
Comment (reidbaker) [inline fanClub/src/com/lastcrusade/fanclub/net/message/FindNewFansMessage.java:10]: So the fact that we get a message is all we need? If that is true should we do the same thing for play, pause, and skip?

**Comment by reidbaker** [comment]
Somewhere there should be a TODO to save the devices we have connected to in the past

File: fanClub/src/com/lastcrusade/fanclub/net/MessageThreadMessageDispatch.java (hunk 14 above)
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/net/MessageThreadMessageDispatch.java:41]: I'll add a few more ? and a couple .
And a comment :laughing:

File: fanClub/src/com/lastcrusade/fanclub/net/message/FindNewFansMessage.java (hunk 15 above)
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/net/message/FindNewFansMessage.java:10]: Correct.  Greenie and I talked about this yesterday, after he had his messages written, and I suggested they could be left alone but future messages can be done this way.

File: fanClub/src/com/lastcrusade/fanclub/net/BluetoothDiscoveryHandler.java (hunk 4 above)
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/net/BluetoothDiscoveryHandler.java:74]: I'm going to say probably not...currently they're the same format, but this is debug output and that is for user presentation.  Those output formats may diverge in the future, and we'd have to change this back.

File: fanClub/lint.xml (hunk 6 above)
Comment (theJenix) [inline fanClub/lint.xml]: I'm not entirely sure...Eclipse generated it.  Lint is a static code checker, so maybe this has something to do with packaging the APK or something.  I think we can safely remove/ignore it.

Comment (theJenix) [inline fanClub/lint.xml]: It has been safely removed/ignored.

File: fanClub/src/com/lastcrusade/fanclub/FanActivity.java (hunk 7 above)
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/FanActivity.java:97]: It's on line 134.

File: fanClub/src/com/lastcrusade/fanclub/FanActivity.java (hunk 8 above)
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/FanActivity.java:112]: This is for handling the string message, which shows "Hello Cleveland" or "You Rock"....I think the Fan gets Hello Cleveland.  It's ultimately a test mechanism, but one I'd like to leave in for now.

File: fanClub/src/com/lastcrusade/fanclub/HostActivity.java (hunk 9 above)
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/HostActivity.java:132]: I added a comment to IMessageHandler to explain the parameters.  It's the bluetooth MAC address.

File: fanClub/src/com/lastcrusade/fanclub/HostActivity.java (hunk 10 above)
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/HostActivity.java]: Fixed.

File: fanClub/src/com/lastcrusade/fanclub/HostActivity.java (hunk 11 above)
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/HostActivity.java]: I removed this check altogether...this method is called to respond to a network message, and onCreate is written to abort if there is no bluetooth, which means that at this point we can assume adapter is nonnull.

File: fanClub/src/com/lastcrusade/fanclub/HostActivity.java (hunk 12 above)
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/HostActivity.java]: Changed to a Log.wtf...this was for testing.

File: fanClub/src/com/lastcrusade/fanclub/net/MessageThreadMessageDispatch.java (hunk 13 above)
Comment (theJenix) [inline fanClub/src/com/lastcrusade/fanclub/net/MessageThreadMessageDispatch.java]: Fixed

**Comment by theJenix** [comment]
I believe I addressed everything in the review.

File: fanClub/src/com/lastcrusade/fanclub/net/message/FindNewFansMessage.java (hunk 15 above)
Comment (dgreenhalgh) [inline fanClub/src/com/lastcrusade/fanclub/net/message/FindNewFansMessage.java:10]: Yeah, I'll look into this. :horse:

**Comment by dgreenhalgh** [comment]
I agree. Does anyone have a problem with discovery being pulled in?

**Comment by reidbaker** [comment]
I Dont let's pull it in

**Pull request merged**